[distributed]:  Nodes fails to restart during a failed chaos test.

<!--- Provide a general summary of the issue in the Title above -->
Chaos test fails for `exec-concurrent TestMakeBucketError` load at a CONCURRENCY level of 100 (Different from the test used in https://github.com/minio/minio/issues/3191).
But All nodes fails to restart with error, more description will be provided below.
## Expected Behavior
<!--- If you're describing a bug, tell us what should happen -->
<!--- If you're suggesting a change/improvement, tell us how it should work -->
- Test to pass.
- Nodes to restart gracefully in Chaos test.
## Current Behavior
<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->
- Firstly the Chaos test under `exec-concurrent TestMakeBucketRegions`(with Concurrency level of 100) load fails. 99 of them pass and only one of them fail.
 
   ```sh
   Goroutine: 51
  FAIL: TestMakeBucketRegions (19.42s)
        api_functional_v4_test.go:138: Error: Multiple disks failures, unable to write data.   6kcyuyx4gjfuki34byeybgh2vimvg0
  FAIL
    ```
- Then multiple nodes fail to restart. 7 out of 8 nodes were subjected to chaos test. All 7 failed to start again. Here are the errors with which the nodes failed to restart.

| Node Number | The error with which the node failed to restart                                                                                         |
|-------------|-----------------------------------------------------------------------------------------------------------------------------------------|
| Node2       | time="2016-11-07T05:24:03Z" level=fatal msg="Unable to load all bucket policies." cause="Bucket not found: lszh0yri6lxneyb2tbaypmmlcw4  |
| Node3       | time="2016-11-07T05:24:03Z" level=fatal msg="Unable to initialize event notification." cause="disk not found" location="[routers.go:74: |
| Node4       | time="2016-11-07T05:24:03Z" level=fatal msg="Unable to load all bucket policies." cause="Bucket not found: lszh0yri6lxneyb2tbaypmmlcw4  |
| Node5       | time="2016-11-07T05:24:03Z" level=fatal msg="Unable to initialize event notification." cause="disk not found" location="[routers.go:74  |
| Node 6      | time="2016-11-07T06:44:24Z" level=fatal msg="Unable to load all bucket policies." cause="Bucket not found: 03rxe6qus9tsmqym1wryg9s09yy  |
| Node8       | time="2016-11-07T05:26:09Z" level=fatal msg="Unable to load all bucket policies." cause="Bucket not found: e31buocts6ti69ixyoid20atlfh  |
| Node 7      | Couldn't find the error in the service status.                                                                                          |
## Possible Solution
<!--- Not obligatory, but suggest a fix/reason for the bug, -->
<!--- or ideas how to implement the addition or change -->
Not sure.
## Steps to Reproduce (for bugs)
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
1. Run [Chaos test](https://github.com/minio/chaos). 
  ```sh
   ./master --endpoints="minio2:9997,minio3:9997,minio4:9997,minio5:9997,minio6:9997,minio7:9997,minio8:9997" -recover=1  -rounds=10 -minio-port=9199
   ```
2. Run [exec-concurrent TestMakeBucketError](https://github.com/minio/perftest/tree/master/exec-concurrent) on the node1 (other 7 nodes are under round robin chaos test) at a CONCURRENCY of 100.
    
3. check  output.log . 

4. Go to each node and check `service minio status`

## Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->
- Chaos test with `exec-concurrent TestMakeBucketError` load at a concurrency of 100.
## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* Version used: Latest master at 9161016962afc92cf42a2dd3cbdbc05aca8ae375 . 
* Server type and version: Packet linux nodes.

- Last time when the issue was witnessed all the node failed to restart. 
- But with the latest master at only 1 node (`node2`) failed with the error below, 
  
  ``` sh
  time="2016-11-14T07:40:09Z" level=fatal msg="Unable to load all bucket policies." cause="Bucket not found: whexylf6qlhh40nmpn5maloykagm . 
  ```
- It has become difficult to reproduce this, had to run the `exec-concurrent TestMakeBucketError` at a concurrency of 50, tens of times (while chaos test is going on) to reproduce. 
- Around like 10th run of the test of one one of the `exec-concurrent` test failed (50 was the concurrency level, so only 1 out of 50 failed)  with following error, 
  
  ``` sh
  
  Goroutine: 21
  --- FAIL: TestMakeBucketError (0.01s)
      api_functional_v4_test.go:92: Error: Your previous request to create the named bucket succeeded and you already own it. ud5tjjk61152vdx1tsn1bs33yggzz9
  FAIL
  ```
- After this happened 1 of the 7 nodes under chaos test failed to restart.

--- a/cmd/bucket-notification-handlers.go
+++ b/cmd/bucket-notification-handlers.go
@@ -54,6 +54,14 @@ func (api objectAPIHandlers) GetBucketNotificationHandler(w http.ResponseWriter,
 	}
 	vars := mux.Vars(r)
 	bucket := vars["bucket"]
+
+	_, err := objAPI.GetBucketInfo(bucket)
+	if err != nil {
+		errorIf(err, "Unable to find bucket info.")
+		writeErrorResponse(w, r, toAPIErrorCode(err), r.URL.Path)
+		return
+	}
+
 	// Attempt to successfully load notification config.
 	nConfig, err := loadNotificationConfig(bucket, objAPI)
 	if err != nil && err != errNoSuchNotifications {
@@ -177,8 +185,7 @@ func PutBucketNotificationConfig(bucket string, ncfg *notificationConfig, objAPI
 		return fmt.Errorf("Unable to persist Bucket notification config to object layer - config=%v errMsg=%v", *ncfg, err)
 	}
 
-	// All servers (including local) are told to update in-memory
-	// config
+	// All servers (including local) are told to update in-memory config
 	S3PeersUpdateBucketNotification(bucket, ncfg)
 
 	return nil

--- a/cmd/bucket-policy-handlers.go
+++ b/cmd/bucket-policy-handlers.go
@@ -142,6 +142,14 @@ func (api objectAPIHandlers) PutBucketPolicyHandler(w http.ResponseWriter, r *ht
 	vars := mux.Vars(r)
 	bucket := vars["bucket"]
 
+	// Before proceeding validate if bucket exists.
+	_, err := objAPI.GetBucketInfo(bucket)
+	if err != nil {
+		errorIf(err, "Unable to find bucket info.")
+		writeErrorResponse(w, r, toAPIErrorCode(err), r.URL.Path)
+		return
+	}
+
 	// If Content-Length is unknown or zero, deny the
 	// request. PutBucketPolicy always needs a Content-Length if
 	// incoming request is not chunked.
@@ -200,13 +208,6 @@ func (api objectAPIHandlers) PutBucketPolicyHandler(w http.ResponseWriter, r *ht
 // persists it to storage, and notify nodes in the cluster about the
 // change. In-memory state is updated in response to the notification.
 func persistAndNotifyBucketPolicyChange(bucket string, pCh policyChange, objAPI ObjectLayer) error {
-	// Verify if bucket actually exists. FIXME: Ideally this check
-	// should not be used but is kept here to error out for
-	// invalid and non-existent buckets.
-	if err := isBucketExist(bucket, objAPI); err != nil {
-		return err
-	}
-
 	// Acquire a write lock on bucket before modifying its
 	// configuration.
 	bucketLock := nsMutex.NewNSLock(bucket, "")
@@ -253,12 +254,18 @@ func (api objectAPIHandlers) DeleteBucketPolicyHandler(w http.ResponseWriter, r 
 	vars := mux.Vars(r)
 	bucket := vars["bucket"]
 
+	// Before proceeding validate if bucket exists.
+	_, err := objAPI.GetBucketInfo(bucket)
+	if err != nil {
+		errorIf(err, "Unable to find bucket info.")
+		writeErrorResponse(w, r, toAPIErrorCode(err), r.URL.Path)
+		return
+	}
+
 	// Delete bucket access policy, by passing an empty policy
 	// struct.
 	if err := persistAndNotifyBucketPolicyChange(bucket, policyChange{true, nil}, objAPI); err != nil {
 		switch err.(type) {
-		case BucketNameInvalid:
-			writeErrorResponse(w, r, ErrInvalidBucketName, r.URL.Path)
 		case BucketPolicyNotFound:
 			writeErrorResponse(w, r, ErrNoSuchBucketPolicy, r.URL.Path)
 		default:
@@ -292,13 +299,19 @@ func (api objectAPIHandlers) GetBucketPolicyHandler(w http.ResponseWriter, r *ht
 	vars := mux.Vars(r)
 	bucket := vars["bucket"]
 
+	// Before proceeding validate if bucket exists.
+	_, err := objAPI.GetBucketInfo(bucket)
+	if err != nil {
+		errorIf(err, "Unable to find bucket info.")
+		writeErrorResponse(w, r, toAPIErrorCode(err), r.URL.Path)
+		return
+	}
+
 	// Read bucket access policy.
 	policy, err := readBucketPolicy(bucket, objAPI)
 	if err != nil {
 		errorIf(err, "Unable to read bucket policy.")
 		switch err.(type) {
-		case BucketNameInvalid:
-			writeErrorResponse(w, r, ErrInvalidBucketName, r.URL.Path)
 		case BucketPolicyNotFound:
 			writeErrorResponse(w, r, ErrNoSuchBucketPolicy, r.URL.Path)
 		default:

--- a/cmd/bucket-policy-handlers_test.go
+++ b/cmd/bucket-policy-handlers_test.go
@@ -250,6 +250,11 @@ func testPutBucketPolicyHandler(obj ObjectLayer, instanceType, bucketName string
 	credentials credential, t *testing.T) {
 	initBucketPolicies(obj)
 
+	bucketName1 := fmt.Sprintf("%s-1", bucketName)
+	if err := obj.MakeBucket(bucketName1); err != nil {
+		t.Fatal(err)
+	}
+
 	// template for constructing HTTP request body for PUT bucket policy.
 	bucketPolicyTemplate := `{"Version":"2012-10-17","Statement":[{"Sid":"","Effect":"Allow","Principal":{"AWS":["*"]},"Action":["s3:GetBucketLocation","s3:ListBucket"],"Resource":["arn:aws:s3:::%s"]},{"Sid":"","Effect":"Allow","Principal":{"AWS":["*"]},"Action":["s3:GetObject"],"Resource":["arn:aws:s3:::%s/this*"]}]}`
 
@@ -327,7 +332,7 @@ func testPutBucketPolicyHandler(obj ObjectLayer, instanceType, bucketName string
 		// setting an invalid bucket policy.
 		// the bucket policy parser will fail.
 		{
-			bucketName:         "non-existent-bucket",
+			bucketName:         bucketName,
 			bucketPolicyReader: bytes.NewReader([]byte("dummy-policy")),
 
 			policyLen:          len([]byte("dummy-policy")),
@@ -339,7 +344,7 @@ func testPutBucketPolicyHandler(obj ObjectLayer, instanceType, bucketName string
 		// Different bucket name used in the HTTP request and the policy string.
 		// checkBucketPolicyResources should fail.
 		{
-			bucketName:         "different-bucket",
+			bucketName:         bucketName1,
 			bucketPolicyReader: bytes.NewReader([]byte(fmt.Sprintf(bucketPolicyTemplate, bucketName, bucketName))),
 
 			policyLen:          len(fmt.Sprintf(bucketPolicyTemplate, bucketName, bucketName)),
@@ -358,7 +363,7 @@ func testPutBucketPolicyHandler(obj ObjectLayer, instanceType, bucketName string
 			policyLen:          len(fmt.Sprintf(bucketPolicyTemplate, bucketName, bucketName)),
 			accessKey:          credentials.AccessKeyID,
 			secretKey:          credentials.SecretAccessKey,
-			expectedRespStatus: http.StatusInternalServerError,
+			expectedRespStatus: http.StatusNotFound,
 		},
 		// Test case - 9.
 		// invalid bucket name is used.
@@ -527,7 +532,7 @@ func testGetBucketPolicyHandler(obj ObjectLayer, instanceType, bucketName string
 			accessKey:            credentials.AccessKeyID,
 			secretKey:            credentials.SecretAccessKey,
 			expectedBucketPolicy: bucketPolicyTemplate,
-			expectedRespStatus:   http.StatusInternalServerError,
+			expectedRespStatus:   http.StatusNotFound,
 		},
 		// Test case - 3.
 		// Case with invalid bucket name.
@@ -736,7 +741,7 @@ func testDeleteBucketPolicyHandler(obj ObjectLayer, instanceType, bucketName str
 			bucketName:         "non-existent-bucket",
 			accessKey:          credentials.AccessKeyID,
 			secretKey:          credentials.SecretAccessKey,
-			expectedRespStatus: http.StatusInternalServerError,
+			expectedRespStatus: http.StatusNotFound,
 		},
 		// Test case - 3.
 		// Case with invalid bucket name.

--- a/cmd/bucket-policy.go
+++ b/cmd/bucket-policy.go
@@ -142,11 +142,6 @@ func getOldBucketsConfigPath() (string, error) {
 // readBucketPolicyJSON - reads bucket policy for an input bucket, returns BucketPolicyNotFound
 // if bucket policy is not found.
 func readBucketPolicyJSON(bucket string, objAPI ObjectLayer) (bucketPolicyReader io.Reader, err error) {
-	// Verify if bucket actually exists
-	if err = isBucketExist(bucket, objAPI); err != nil {
-		return nil, err
-	}
-
 	policyPath := pathJoin(bucketConfigPrefix, bucket, policyJSON)
 	objInfo, err := objAPI.GetObjectInfo(minioMetaBucket, policyPath)
 	err = errorCause(err)

--- a/cmd/event-notifier.go
+++ b/cmd/event-notifier.go
@@ -466,7 +466,6 @@ func loadAllBucketNotifications(objAPI ObjectLayer) (map[string]*notificationCon
 		} else {
 			nConfigs[bucket.Name] = nCfg
 		}
-
 		lCfg, lErr := loadListenerConfig(bucket.Name, objAPI)
 		if lErr != nil {
 			if lErr != errNoSuchNotifications {

--- a/cmd/web-handlers_test.go
+++ b/cmd/web-handlers_test.go
@@ -1118,8 +1118,6 @@ func testWebSetBucketPolicyHandler(obj ObjectLayer, instanceType string, t TestE
 		policy     string
 		pass       bool
 	}{
-		// Inexistent bucket
-		{"fooo", "", "readonly", false},
 		// Invalid bucket name
 		{"", "", "readonly", false},
 		// Invalid policy

